File: 2023/day05/node/solution.js
import { readFileSync } from 'fs';
import { dirname, join } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));

/**
 * @typedef {Object} MappingRange
 * @property {number} destinationStart - Start of the destination range
 * @property {number} sourceStart - Start of the source range
 * @property {number} rangeLength - Length of the range
 */

/**
 * @typedef {Object} AlmanacData
 * @property {number[]} seeds - List of seed numbers
 * @property {MappingRange[][]} categoryMaps - Array of category mappings (seed-to-soil, soil-to-fertilizer, etc.)
 */

/**
 * @typedef {[number, number]} SeedRange
 * A tuple representing [startInclusive, endExclusive] of a seed range
 */

/**
 * Parses the almanac input text into structured data.
 *
 * @param {string} inputText - Raw input text from the puzzle
 * @returns {AlmanacData} Parsed seeds and category mappings
 */
function parseInput(inputText) {
  const sections = inputText.trim().split('\n\n');

  const seeds = sections[0]
    .split(': ')[1]
    .split(' ')
    .map(Number);

  const categoryMaps = sections.slice(1).map((section) => {
    const lines = section.trim().split('\n');
    return lines.slice(1).map((line) => {
      const [destinationStart, sourceStart, rangeLength] = line.split(' ').map(Number);
      return { destinationStart, sourceStart, rangeLength };
    });
  });

  return { seeds, categoryMaps };
}

/**
 * Applies a single category map to transform a source value to a destination value.
 * If no mapping range contains the value, returns the value unchanged (identity mapping).
 *
 * @param {number} sourceValue - The value to transform
 * @param {MappingRange[]} mappingRanges - Array of mapping ranges for this category
 * @returns {number} The transformed destination value
 */
function applyCategoryMap(sourceValue, mappingRanges) {
  for (const { destinationStart, sourceStart, rangeLength } of mappingRanges) {
    const sourceEnd = sourceStart + rangeLength;
    if (sourceValue >= sourceStart && sourceValue < sourceEnd) {
      const offset = sourceValue - sourceStart;
      return destinationStart + offset;
    }
  }
  return sourceValue;
}

/**
 * Converts a seed number to its final location by applying all category mappings
 * in sequence (seed -> soil -> fertilizer -> water -> light -> temperature -> humidity -> location).
 *
 * @param {number} seedNumber - The initial seed number
 * @param {MappingRange[][]} categoryMaps - All category mappings in order
 * @returns {number} The final location number
 */
function seedToLocation(seedNumber, categoryMaps) {
  let currentValue = seedNumber;
  for (const mappingRanges of categoryMaps) {
    currentValue = applyCategoryMap(currentValue, mappingRanges);
  }
  return currentValue;
}

/**
 * Solves Part 1: Find the lowest location number for any of the individual seed numbers.
 *
 * @param {number[]} seeds - List of individual seed numbers
 * @param {MappingRange[][]} categoryMaps - All category mappings
 * @returns {number} The minimum location number
 */
function part1(seeds, categoryMaps) {
  const locationNumbers = seeds.map((seed) => seedToLocation(seed, categoryMaps));
  return Math.min(...locationNumbers);
}

/**
 * Applies a category map to a collection of seed ranges, transforming them to destination ranges.
 * Handles range splitting when source ranges partially overlap with mapping ranges.
 *
 * @param {SeedRange[]} inputRanges - Array of [start, end) ranges to transform
 * @param {MappingRange[]} mappingRanges - Mapping ranges for this category
 * @returns {SeedRange[]} Transformed destination ranges
 */
function applyCategoryMapToRanges(inputRanges, mappingRanges) {
  const transformedRanges = [];

  for (const [rangeStart, rangeEnd] of inputRanges) {
    let unmappedSegments = [[rangeStart, rangeEnd]];

    for (const { destinationStart, sourceStart, rangeLength } of mappingRanges) {
      const sourceEnd = sourceStart + rangeLength;
      const mappingOffset = destinationStart - sourceStart;
      const nextUnmappedSegments = [];

      for (const [segmentStart, segmentEnd] of unmappedSegments) {
        // Segment portion before the mapping source range (remains unmapped)
        if (segmentStart < sourceStart) {
          const beforeEnd = Math.min(segmentEnd, sourceStart);
          nextUnmappedSegments.push([segmentStart, beforeEnd]);
        }

        // Segment portion overlapping with the mapping source range (gets transformed)
        const overlapStart = Math.max(segmentStart, sourceStart);
        const overlapEnd = Math.min(segmentEnd, sourceEnd);
        if (overlapStart < overlapEnd) {
          transformedRanges.push([
            overlapStart + mappingOffset,
            overlapEnd + mappingOffset,
          ]);
        }

        // Segment portion after the mapping source range (remains unmapped)
        if (segmentEnd > sourceEnd) {
          const afterStart = Math.max(segmentStart, sourceEnd);
          nextUnmappedSegments.push([afterStart, segmentEnd]);
        }
      }

      unmappedSegments = nextUnmappedSegments;
    }

    // Any segments that didn't match any mapping range use identity mapping
    transformedRanges.push(...unmappedSegments);
  }

  return transformedRanges;
}

/**
 * Solves Part 2: Find the lowest location number where seeds are interpreted as ranges.
 * Each pair of numbers represents (start, length) of a seed range.
 *
 * @param {number[]} seeds - Seed data as pairs of [start, length]
 * @param {MappingRange[][]} categoryMaps - All category mappings
 * @returns {number} The minimum location number from any seed in any range
 */
function part2(seeds, categoryMaps) {
  // Convert seed pairs to [start, end) ranges
  const seedRanges = [];
  for (let i = 0; i < seeds.length; i += 2) {
    const rangeStart = seeds[i];
    const rangeLength = seeds[i + 1];
    seedRanges.push([rangeStart, rangeStart + rangeLength]);
  }

  // Apply each category map to transform all ranges through the pipeline
  let currentRanges = seedRanges;
  for (const mappingRanges of categoryMaps) {
    currentRanges = applyCategoryMapToRanges(currentRanges, mappingRanges);
  }

  // Find the minimum starting value across all final location ranges
  const rangeStarts = currentRanges.map(([start]) => start);
  return Math.min(...rangeStarts);
}

// Main execution
const inputPath = join(__dirname, '..', 'input.txt');
const inputText = readFileSync(inputPath, 'utf-8');

const { seeds, categoryMaps } = parseInput(inputText);

console.log('Part 1:', part1(seeds, categoryMaps));
console.log('Part 2:', part2(seeds, categoryMaps));

File: 2023/day05/rust/solution.rs
use std::fs;
use std::ops::Range;

/// Represents a mapping rule that transforms source values to destination values.
///
/// The mapping applies to values in the range `source_start..(source_start + length)`,
/// shifting them by `(destination_start - source_start)`.
#[derive(Debug, Clone)]
struct MappingRule {
    destination_start: u64,
    source_range: Range<u64>,
}

impl MappingRule {
    /// Creates a new mapping rule from destination start, source start, and length.
    fn new(destination_start: u64, source_start: u64, length: u64) -> Self {
        Self {
            destination_start,
            source_range: source_start..(source_start + length),
        }
    }

    /// Returns the offset to apply when mapping values within this rule's range.
    fn offset(&self) -> i64 {
        self.destination_start as i64 - self.source_range.start as i64
    }

    /// Checks if a value falls within this mapping rule's source range.
    fn contains(&self, value: u64) -> bool {
        self.source_range.contains(&value)
    }

    /// Applies this mapping rule to a value, returning the mapped value.
    fn apply(&self, value: u64) -> u64 {
        (value as i64 + self.offset()) as u64
    }
}

/// A category map containing multiple mapping rules (e.g., seed-to-soil).
type CategoryMap = Vec<MappingRule>;

/// An interval representing a range of values [start, end).
#[derive(Debug, Clone, Copy)]
struct Interval {
    start: u64,
    end: u64,
}

impl Interval {
    /// Creates a new interval from start (inclusive) to end (exclusive).
    fn new(start: u64, end: u64) -> Self {
        Self { start, end }
    }

    /// Checks if the interval is non-empty.
    fn is_valid(&self) -> bool {
        self.start < self.end
    }

    /// Returns the intersection of this interval with another range.
    fn intersect(&self, range: &Range<u64>) -> Option<Self> {
        let overlap_start = self.start.max(range.start);
        let overlap_end = self.end.min(range.end);
        if overlap_start < overlap_end {
            Some(Self::new(overlap_start, overlap_end))
        } else {
            None
        }
    }
}

/// Parses the puzzle input into seeds and a list of category maps.
///
/// # Returns
/// A tuple containing:
/// - A vector of seed numbers
/// - A vector of category maps, each containing mapping rules
fn parse_input(input_text: &str) -> (Vec<u64>, Vec<CategoryMap>) {
    let sections: Vec<&str> = input_text.trim().split("\n\n").collect();

    let seeds = sections[0]
        .split(": ")
        .nth(1)
        .expect("Input should contain seeds after 'seeds: '")
        .split_whitespace()
        .filter_map(|s| s.parse().ok())
        .collect();

    let category_maps = sections[1..]
        .iter()
        .map(|section| {
            section
                .lines()
                .skip(1) // Skip the header line (e.g., "seed-to-soil map:")
                .filter_map(|line| {
                    let numbers: Vec<u64> = line
                        .split_whitespace()
                        .filter_map(|s| s.parse().ok())
                        .collect();
                    if numbers.len() == 3 {
                        Some(MappingRule::new(numbers[0], numbers[1], numbers[2]))
                    } else {
                        None
                    }
                })
                .collect()
        })
        .collect();

    (seeds, category_maps)
}

/// Applies a single category map to transform a value.
///
/// If the value falls within any mapping rule's source range, it is transformed.
/// Otherwise, the value passes through unchanged.
fn apply_category_map(value: u64, category_map: &CategoryMap) -> u64 {
    category_map
        .iter()
        .find(|rule| rule.contains(value))
        .map(|rule| rule.apply(value))
        .unwrap_or(value)
}

/// Converts a seed number to a location number by applying all category maps.
fn seed_to_location(seed: u64, category_maps: &[CategoryMap]) -> u64 {
    category_maps
        .iter()
        .fold(seed, |current_value, category_map| {
            apply_category_map(current_value, category_map)
        })
}

/// Finds the lowest location number for any initial seed (Part 1).
///
/// Each seed number is transformed through all category maps, and the
/// minimum resulting location is returned.
fn solve_part1(seeds: &[u64], category_maps: &[CategoryMap]) -> u64 {
    seeds
        .iter()
        .map(|&seed| seed_to_location(seed, category_maps))
        .min()
        .expect("At least one seed should be present")
}

/// Applies a category map to a list of intervals, returning transformed intervals.
///
/// Each input interval may be split into multiple output intervals based on
/// how it overlaps with the mapping rules.
fn apply_category_map_to_intervals(
    input_intervals: &[Interval],
    category_map: &CategoryMap,
) -> Vec<Interval> {
    input_intervals
        .iter()
        .flat_map(|&interval| {
            let mut mapped_intervals = Vec::new();
            let mut unmapped_intervals = vec![interval];

            for rule in category_map {
                let mut still_unmapped = Vec::new();

                for current in unmapped_intervals {
                    // Part before the mapping rule's range (stays unmapped)
                    let before = Interval::new(current.start, current.end.min(rule.source_range.start));
                    if before.is_valid() {
                        still_unmapped.push(before);
                    }

                    // Part overlapping with the mapping rule's range (gets mapped)
                    if let Some(overlap) = current.intersect(&rule.source_range) {
                        mapped_intervals.push(Interval::new(
                            rule.apply(overlap.start),
                            rule.apply(overlap.end),
                        ));
                    }

                    // Part after the mapping rule's range (stays unmapped)
                    let after = Interval::new(current.start.max(rule.source_range.end), current.end);
                    if after.is_valid() {
                        still_unmapped.push(after);
                    }
                }

                unmapped_intervals = still_unmapped;
            }

            // Remaining unmapped intervals pass through unchanged
            mapped_intervals.extend(unmapped_intervals);
            mapped_intervals
        })
        .collect()
}

/// Finds the lowest location for seed ranges (Part 2).
///
/// Seeds are interpreted as pairs: (start, length), defining ranges of seeds.
/// The solution processes these ranges efficiently without iterating over
/// individual seed values.
fn solve_part2(seeds: &[u64], category_maps: &[CategoryMap]) -> u64 {
    let initial_intervals: Vec<Interval> = seeds
        .chunks_exact(2)
        .map(|chunk| Interval::new(chunk[0], chunk[0] + chunk[1]))
        .collect();

    let final_intervals = category_maps
        .iter()
        .fold(initial_intervals, |intervals, category_map| {
            apply_category_map_to_intervals(&intervals, category_map)
        });

    final_intervals
        .iter()
        .map(|interval| interval.start)
        .min()
        .expect("At least one interval should be present")
}

fn main() {
    let input_path = std::path::Path::new(file!())
        .parent()
        .unwrap()
        .join("../input.txt");
    let input_text = fs::read_to_string(&input_path)
        .unwrap_or_else(|_| fs::read_to_string("../input.txt").unwrap());

    let (seeds, category_maps) = parse_input(&input_text);

    println!("Part 1: {}", solve_part1(&seeds, &category_maps));
    println!("Part 2: {}", solve_part2(&seeds, &category_maps));
}
